### PR TITLE
make shardingcodec pickleable 

### LINKED ISSUE #1990: ShardingCodec cannot be pickled
Because it uses [`lru_cache`](https://github.com/zarr-developers/zarr-python/blob/e3ee09eaa8a5e17a92d00ae4979aa069f755eaba/src/zarr/codecs/sharding.py#L323-L325) and we have not implemented the requisite dunder methods for object serialization to make it pickleable, which is not hard according to [internet wisdom](https://stackoverflow.com/a/73785520) .

todo: 
- add tests for pickleability for the ingredients of `Array` and `Group` classes
- implement `__getstate__` and `__setstate__` as needed to make failing tests pass

### PR DESCRIPTION
**Pull request by d-v-b**
- Adds dunder methods to `ShardingCodec` to enable pickling, and adds a simple test for this
 - refactors the codecs test suite. this is the start of a larger effort.
 - fixes a small bug in morton order calculation that broke sharding for 1D arrays, and adds a test for it
- removes some tests from `test_codecs` that were not testing codecs, and were instead testing things that we are testing elsewhere. I suspect these tests were copied in from zarrita.

closes #1990 

TODO:
* [ ] Add unit tests and/or doctests in docstrings
* [ ] Add docstrings and API docs for any new/modified user-facing classes and functions
* [ ] New/modified features documented in docs/tutorial.rst
* [ ] Changes documented in docs/release.rst
* [ ] GitHub Actions have all passed
* [ ] Test coverage is 100% (Codecov passes)

### CODE REVIEW

File: tests/v3/test_codecs/test_endian.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        chunk_key_encoding=("v2", "."),
+        codecs=[BytesCodec(endian=endian)],
+    )
+
+    await _AsyncArrayProxy(a)[:, :].set(data)
+    readback_data = await _AsyncArrayProxy(a)[:, :].get()
+    assert np.array_equal(data, readback_data)
+
+    # Compare with zarr-python
+    z = zarr.v2.create(
```
Comment (normanrz) [inline tests/v3/test_codecs/test_endian.py:38]: We probably need to change these before we get rid of the v2 package.

**Review by normanrz — APPROVED** [review]
Thanks! I love the work on the tests.

File: tests/v3/test_codecs/test_endian.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        fill_value=0,
+        chunk_key_encoding=("v2", "."),
+        codecs=[BytesCodec(endian=endian)],
+    )
+
+    await _AsyncArrayProxy(a)[:, :].set(data)
+    readback_data = await _AsyncArrayProxy(a)[:, :].get()
+    assert np.array_equal(data, readback_data)
+
+    # Compare with zarr-python
```
Comment (d-v-b) [inline tests/v3/test_codecs/test_endian.py:37]:
```suggestion
    # Compare with v2
```

File: tests/v3/test_codecs/test_endian.py (hunk 1 above)
Comment (d-v-b) [reply tests/v3/test_codecs/test_endian.py:38]: i changed "zarr-python" to "v2", so it looks a little less odd

**Pull request merged**